application-task-default/src/main/java/com/xwiki/task/internal/notifications/TaskChangedEventNotificationListener.java

```diff
@@ -0,0 +1,218 @@
+        } else if (property instanceof DateProperty) {
+            return (Date) ((DateProperty) property).getValue();
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Get a {@link TaskChangedEvent} reflecting the changes made to a certain property, if a change happened.
+     * 
+     * @param currentObject the current version of the object
+     * @param previousObject the previous version of the object
+     * @param propertyName the property to check for changes
+     * @param baseEvent a skeleton event used as a template when calling multiple times for the same object
+     * @return the event describing the changes done to the specified field, null when no change occured
+     */
+    private TaskChangedEvent getFieldChangedEvent(BaseObject currentObject, BaseObject previousObject,
+        String propertyName, TaskChangedEvent baseEvent)
+    {
+        Object currentValue = getPropertyValue(currentObject, propertyName);
+        Object previousValue = getPropertyValue(previousObject, propertyName);
+
+        if (currentValue == null) {
+            return null;
+        }
+
+        boolean valuesAreEqual = currentValue.equals(previousValue);
+
+        String localizationSuffix = propertyName;
+
+        if (currentValue instanceof Date) {
+            long currentDate = ((Date) currentValue).getTime();
+            long previousDate = ((Date) previousValue).getTime();
+
+            valuesAreEqual = (currentDate == previousDate);
+            if (currentDate > previousDate) {
+                localizationSuffix += ".later";
+            } else {
+                localizationSuffix += ".earlier";
+            }
+        }
+
+        if (valuesAreEqual) {
+            return null;
+        }
+
+        Map<String, Object> localizationParams = new HashMap<String, Object>();
+        localizationParams.put("new", currentValue);
+        localizationParams.put("old", previousValue);
+        localizationParams.put("type", localizationSuffix);
+
+        TaskChangedEvent event = baseEvent.clone();
+        event.setLocalizationParams(localizationParams);
+
+        return event;
+    }
+
+    @Override
+    public void onEvent(Event event, Object source, Object data)
+    {
+        if (event instanceof XObjectUpdatedEvent) {
+            onEvent((XObjectUpdatedEvent) event, (XWikiDocument) source, (XWikiContext) data);
+        } else if (event instanceof XObjectAddedEvent) {
+            onEvent((XObjectAddedEvent) event, (XWikiDocument) source, (XWikiContext) data);
+        }
+    }
+
+    private void onEvent(XObjectUpdatedEvent event, XWikiDocument source, XWikiContext data)
+    {
+        XWikiContext context = (XWikiContext) data;
+
+        BaseObject currentObject = source.getXObject(source.resolveClassReference(TASK_MANAGER_CLASS_NAME));
+        BaseObject previousObject = context.getDoc().getXObject(source.resolveClassReference(TASK_MANAGER_CLASS_NAME));
```

Comment:
This is still wrong, as editing a task from another page (like from the Gantt diagram on the TaskManager homepage) will try to fetch the object from that page, not the actual task page.

Reply:
I was checking the [XObjectUpdatedEvent](https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/event/XObjectUpdatedEvent.java#L30) class documentation and it states that the source parameter should also the 'original' document
You could check [this](https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-notifiers/xwiki-platform-notifications-notifiers-api/src/main/java/org/xwiki/notifications/notifiers/internal/email/IntervalUsersManagerInvalidator.java#L98) to see how the original document was accessed